### PR TITLE
Try deducing value type from return type in parallel reduce

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/kokkos/kokkos/issues/5156, improves https://github.com/kokkos/kokkos/issues/2034.

Extending `FunctorAnalysis` to be able to be used with templated call operators appears to be very difficult. On the other hand, the overloads of `parallel_reduce` and `parallel_scan` taking a return value already contain a hint for the `value_type` to use so that we can avoid using the deduction in `FunctorAnalysis` in those cases.

This pull request does exactly that. If the functor has a `value_type` alias, we use that, otherwise, we use the `value_type` deduced from the return type and if that is `void`, too, we do the deduction in `FunctorAnalysis` based on the functor's call operator.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

File: core/unit_test/incremental/Test16_ParallelScan.hpp [hunk 1]
```diff
@@ -57,25 +57,30 @@ const int N      = 10;
...
     // Exclusive parallel_scan call.
-    Kokkos::parallel_scan(
-        Kokkos::RangePolicy<ExecSpace>(0, N),
-        KOKKOS_LAMBDA(const int i, value_type &update_value, const bool final) {
-          const value_type val_i = d_data(i);
-          if (final) d_data(i) = update_value;
-
-          update_value += val_i;
-        });
+    Kokkos::parallel_scan(Kokkos::RangePolicy<ExecSpace>(0, N), *this, total);
```
Comment (PhilMiller) [inline core/unit_test/incremental/Test16_ParallelScan.hpp:83]: It may also be worth testing C++14 generic lambdas, e.g.

```c++
Kokkos::parallel_scan(Kokkos::RangePolicy<ExecSpace>(0, N), KOKKOS_LAMBDA(auto i, value_type &update_value, const bool final) { ...
```

**Review by PhilMiller — APPROVED** [review]
I don't have a good sense of what the base class change in OpenMPTarget is doing. Other than that, this looks good to me. 

Testing lambdas that are generic over the index argument would be nice to add, but I don't see it as necessary.

**Comment by masterleinad** [comment]
> Testing lambdas that are generic over the index argument would be nice to add, but I don't see it as necessary.

I tried that first but `Cuda` doesn't allow generic `__host__ __device__` lambdas.

**Comment by masterleinad** [comment]
This needs more discussion anyway. The main question is if we primarily let the functor or the return value let determine the value type. In the former case, the question is if we allow proxy types as return value with the only requirement that they must be assignable.

**Comment by masterleinad** [comment]
Again, only `OpenACC` is spuriously failing:
```
 Start 33: KokkosCore_IncrementalTest_OPENACC

33: Test command: /var/jenkins/workspace/Kokkos/build/core/unit_test/KokkosCore_IncrementalTest_OPENACC
33: Test timeout computed to be: 1500
33: Error: no GPU available for execution.
33:  Raised by Kokkos::initialize().
33: 
33: Backtrace:
33:                         [0x4b2453]
33:                         [0x490463]
33:                         [0x485cf3]
33:                         [0x485e32]
33:                         [0x4b68d1]
33:                         [0x48e81c]
33:                         [0x48a048]
33:                         [0x412fdd]
33: __libc_start_main [0x7f445cce20b3]
33:                         [0x412e6e]
33/48 Test #33: KokkosCore_IncrementalTest_OPENACC ...........Subprocess aborted***Exception:   0.19 sec
```

File: core/unit_test/incremental/Test16_ParallelScan.hpp (hunk 1 above)
Comment (masterleinad) [reply core/unit_test/incremental/Test16_ParallelScan.hpp:83]: `Cuda` doesn't like that.

**Comment by masterleinad** [comment]
Deducing the value type from the final member function for parallel_reduce without return type takes some more work and I think it's fine to consider doing this in another pull request if we really want.

File: core/unit_test/CMakeLists.txt [hunk 2]
```diff
@@ -476,12 +476,12 @@ if (Kokkos_ENABLE_OPENMP)
...
-    SOURCES
-      UnitTestMainInit.cpp
-      openmp/TestOpenMP_Graph.cpp
-  )
+  #KOKKOS_ADD_EXECUTABLE_AND_TEST(
+  #  UnitTest_OpenMPGraph
+  #  SOURCES
+  #    UnitTestMainInit.cpp
+  #    openmp/TestOpenMP_Graph.cpp
+  #)
```
Comment (ldh4) [inline core/unit_test/CMakeLists.txt:484]: Any reasons for commenting these out instead of deleting them? Are these expected to be enabled again?

File: core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp [hunk 3]
```diff
@@ -704,35 +704,37 @@ __device__ bool cuda_single_inter_block_reduce_scan(
...
       Impl::FunctorAnalysis<Impl::FunctorPatternInterface::REDUCE,
-                            RangePolicy<Cuda, ArgTag>, FunctorType>;
+                            RangePolicy<Cuda, ArgTag>, FunctorType, ValueType>;
 
   return (BlockSize + 2) * Analysis::value_size(functor);
 }
 
-template <typename WorkTag, typename Policy, typename FunctorType>
+template <typename ValueType, typename WorkTag, typename Policy,
+          typename FunctorType>
```
Comment (ldh4) [inline core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp:730]: Can `typename ValueType` be placed at the end of template parameters? Throughout this PR, the new `ValueType` has consistently been added toward the end of the template parameters, but this seems to be the only place where it was placed in front.

**Review by ldh4 — APPROVED** [review]

File: core/unit_test/CMakeLists.txt (hunk 2 above)
Comment (masterleinad) [reply core/unit_test/CMakeLists.txt:484]: Will fix.

File: core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp (hunk 3 above)
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp:730]: Will change. In fact, #5332 would remove the template parameter again but the order of the pull requests doesn't matter much in my opinion.

Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp:730]: Ah, in fact, I prefer leaving it this way so that `Policy` and `FunctorType` can be deduced.

**Review by crtrott — CHANGES_REQUESTED** [review]
This actually did break compilation of the OpenACC backend, needs fixing.

**Comment by masterleinad** [comment]
> This actually did break compilation of the OpenACC backend, needs fixing.

:unamused: :

**Comment by masterleinad** [comment]
@crtrott I fixed OpenACC.

File: core/src/Kokkos_ExecPolicy.hpp [hunk 4]
```diff
@@ -914,7 +914,7 @@ struct PatternImplSpecializationFromTag<Kokkos::ParallelForTag, Args...>
 
 template <class... Args>
 struct PatternImplSpecializationFromTag<Kokkos::ParallelReduceTag, Args...>
-    : identity<ParallelReduce<Args...>> {};
+    : identity<ParallelReduce<Args..., void>> {};
```
Comment (dalg24) [inline core/src/Kokkos_ExecPolicy.hpp:917]: Why did you change this?

Comment (masterleinad) [reply core/src/Kokkos_ExecPolicy.hpp:917]: I didn't bother to try to figure out how to make the detection work for `Graph` kernels, here, and we just pass the original arguments here and force querying the functor for the value type. This is fixed in #5332.
I could look into fixing it here, too, if required/requested.

**Review by PhilMiller — APPROVED** [review]

**Comment by masterleinad** [comment]
Let's see if we can get #5332 in than we don't need this one.

**Comment by masterleinad** [comment]
Supersed by #5976.
